### PR TITLE
Check if command is chained before making API call

### LINKED ISSUE #981: Error when using hyphenated collection names
When I try to perform a find operation in a collection with a hyphenated name I get this error.

## Repro steps:

1. Open a scratchbook
2. `db.getCollection('my-collection').find({})`
3. Execute

## Result:

Action: cosmosDB.executeMongoCommand
Error type: Error
Error Message: collection name must be a String

Version: 0.9.1
OS: darwin

### PR DESCRIPTION
**Pull request by PrashanthCorp**
Fixes #981, #998 
Fix part of #1096.

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -9481,25 +9555,18 @@
 			}
 		},
 		"vscode": {
-			"version": "1.1.26",
-			"resolved": "https://registry.npmjs.org/vscode/-/vscode-1.1.26.tgz",
-			"integrity": "sha512-z1Nf5J38gjUFbuDCbJHPN6OJ//5EG+e/yHlh6ERxj/U9B2Qc3aiHaFr38/fee/GGnxvRw/XegLMOG+UJwKi/Qg==",
+			"version": "1.1.33",
```
Comment (StephenWeatherford) [inline package-lock.json]: Where are these changes coming from?

File: src/mongo/MongoScrapbook.ts [hunk 2]
```diff
@@ -112,16 +112,13 @@ async function executeCommand(activeEditor: vscode.TextEditor, database: MongoDa
...
-		if (command.name === 'find') {
+		if (command.name === 'find' && !!command.collection && !command.chained) {
 			await editorManager.showDocument(new MongoFindResultEditor(database, command), 'cosmos-result.json', { showInNextColumn: true });
 		} else {
 			const result = await database.executeCommand(command, context);
-			if (command.name === 'findOne') {
-				if (result === "null") {
-					throw new Error(`Could not find any documents`);
-				}
+			if (command.name === 'findOne' && !!command.collection) {
```
Comment (StephenWeatherford) [inline src/mongo/MongoScrapbook.ts]: Don't you need that same check for findOne?

File: test/mongoGetCommand.test.ts [hunk 3]
```diff
@@ -934,6 +932,23 @@ suite("scrapbook parsing Tests", () => {
         });
     });
 
+    test("Chained command where final function call is recognized. Ensure execution is sent to shell - test user issues: https://github.com/Microsoft/vscode-cosmosdb/issues/981", () => {
+        testParse(`db.getCollection('my-collection').find({});`, {
```
Comment (StephenWeatherford) [inline test/mongoGetCommand.test.ts]: What about this checks that it gets sent to the shell?

Comment (PrashanthCorp) [reply test/mongoGetCommand.test.ts]: The command is chained - MongoFindResultEditor cannot handle this.

Comment (PrashanthCorp) [reply test/mongoGetCommand.test.ts]: > What about this checks that it gets sent to the shell?

The collection name is undefined. A command like `db.abc.find()` would have the collection name be abc.

File: package-lock.json (hunk 1 above)
Comment (PrashanthCorp) [reply package-lock.json]: I'm assuming they come from an npm install after I merge from master

File: src/mongo/editors/MongoFindOneResultEditor.ts [hunk 4]
```diff
@@ -18,6 +18,9 @@ export class MongoFindOneResultEditor implements ICosmosEditor<IMongoDocument> {
     constructor(databaseNode: MongoDatabaseTreeItem, collectionName: string, data: string) {
         this._databaseNode = databaseNode;
         this._collectionName = collectionName;
+        if (data === "null") {
```
Comment (EricJizbaMSFT) [inline src/mongo/editors/MongoFindOneResultEditor.ts]: I don't understand this check. Is it really checking for the _string_ null? When can this happen?

File: test/mongoGetCommand.test.ts [hunk 5]
```diff
@@ -934,6 +932,23 @@ suite("scrapbook parsing Tests", () => {
...
+    test("Chained command where final function call is recognized. Ensure execution is sent to shell (the command has not collection name) - test user issues: https://github.com/Microsoft/vscode-cosmosdb/issues/981", () => {
+        testParse(`db.getCollection('my-collection').find({});`, {
+            collection: undefined,
+            name: "find",
+            args: ["my-collection", {}]
+        });
+    });
+
+    test("Chained command where final function call is recognized as findOne. Ensure execution is sent to shell (the command has not collection name) - test user issues: https://github.com/Microsoft/vscode-cosmosdb/issues/981", () => {
+        testParse(`db.getCollection('my-collection').find({});`, {
```
Comment (EricJizbaMSFT) [inline test/mongoGetCommand.test.ts]: Is this supposed to be `findOne`? It looks like a dupe of the above test.

File: package-lock.json (hunk 1 above)
Comment (EricJizbaMSFT) [reply package-lock.json]: Either way they shouldn't be a part of a PR where you didn't change "package.json". You either need to merge this branch with master or clean up somehow

File: test/mongoGetCommand.test.ts (hunk 3 above)
Comment (EricJizbaMSFT) [reply test/mongoGetCommand.test.ts]: I agree with Stephen that the name of this test isn't accurate. I would just name it something like `find command for collection with hyphens in name, from #981`

If you want to create a test that represents the name you currently have, it should somehow actually verify against the shell. But I don't think that's worth doing until after https://github.com/microsoft/vscode-cosmosdb/pull/1105

File: src/mongo/editors/MongoFindOneResultEditor.ts (hunk 4 above)
Comment (PrashanthCorp) [reply src/mongo/editors/MongoFindOneResultEditor.ts]: Yes. The variable name `data` is misleading - `data` is a _string_  of the query result (returned from `executeCommand`), and hence line 25 has a `EJSON.parse`. 

It is the string "null" when there are no documents matching the filter. :)

Comment (PrashanthCorp) [reply src/mongo/editors/MongoFindOneResultEditor.ts]: I didn't see the need to rename the variable since the type annotations explain the expected handling.

File: src/mongo/MongoScrapbook.ts [hunk 6]
```diff
@@ -348,7 +345,7 @@ class FindMongoCommandsVisitor extends MongoVisitor<MongoCommand[]> {
 		}
 		let argumentContextArray: mongoParser.ArgumentContext[] = filterType(argumentsToken.children, mongoParser.ArgumentContext);
 
-		let functionMap = { "ObjectId": this.objectIdToObject, "ISODate": this.dateToObject, "Date": this.dateToObject };
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoScrapbook.ts]: Why did this change? This doesn't seem related to https://github.com/microsoft/vscode-cosmosdb/issues/981. If you're fixing a different bug I would prefer a separate PR or at least mention that bug in the PR description.

File: src/mongo/editors/MongoFindOneResultEditor.ts (hunk 4 above)
Comment (EricJizbaMSFT) [reply src/mongo/editors/MongoFindOneResultEditor.ts]: Suggest adding a comment. A string of "null" is very uncommon, plus this repo doesn't have strictNullChecks on so the types aren't super accurate. Someone could easily mess this up if they didn't know the full story. Just add something like:
```
// Intentionally checking against string of "null", which can be returned by mongo shell
```

File: src/mongo/MongoScrapbook.ts [hunk 7]
```diff
@@ -221,7 +218,7 @@ class FindMongoCommandsVisitor extends MongoVisitor<MongoCommand[]> {
 			name: '',
 			arguments: [],
 			argumentObjects: [],
-			chained: funcCallCount > 1 ? true : false
+			sendToShell: funcCallCount > 1 ? true : false
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoScrapbook.ts]: I don't think it's accurate to rename `chained` to `sendToShell`. We use the shell as a catch-all for several different unsupported cases, one of which happens to be `chained`. I would prefer to leave the name as-is and wait for https://github.com/microsoft/vscode-cosmosdb/pull/1105 to do any refactoring related to this

File: src/mongo/MongoScrapbook.ts (hunk 6 above)
Comment (PrashanthCorp) [reply src/mongo/MongoScrapbook.ts]: The changes there would depend on this PR (doing it independently complicates the code).  I chose to address the issue here. I mentioned it in the commit message.  Will add to PR description.

File: src/mongo/MongoScrapbook.ts (hunk 7 above)
Comment (PrashanthCorp) [reply src/mongo/MongoScrapbook.ts]: This is related to the fix for #1096. When both bugs are fixed (chained commands and the Date), having two separate flags for the same action of sending to the shell is redundant when on flag would do.

File: src/mongo/tree/MongoDatabaseTreeItem.ts [hunk 8]
```diff
@@ -97,7 +97,7 @@ export class MongoDatabaseTreeItem extends AzureParentTreeItem<IMongoTreeRoot> {
 	}
 
 	async executeCommand(command: MongoCommand, context: IActionContext): Promise<string> {
-		if (command.collection) {
+		if (command.collection && !command.sendToShell) {
```
Comment (EricJizbaMSFT) [inline src/mongo/tree/MongoDatabaseTreeItem.ts]: If you want to name it `sendToShell` - I think we need to refactor this method. Something like this:
```
if (!command.sendToShell) {
	if (command.collection) {
		let db = await this.getDb();
		const collection = db.collection(command.collection);
		if (collection) {
			const collectionTreeItem = new MongoCollectionTreeItem(this, collection, command.arguments);
			const result = await collectionTreeItem.executeCommand(command.name, command.arguments);
			if (result) {
				return result;
			}
		}
	} else if (command.name === 'createCollection') {
		return withProgress(this.createCollection(stripQuotes(command.arguments.join(','))).then(() => JSON.stringify({ 'Created': 'Ok' })), 'Creating collection');
	}
}

return withProgress(this.executeCommandInShell(command, context), 'Executing command');
```

My problem with the current name is that there are cases where sendToShell could be `true`, but it's not actually sent to the shell. With my above code, that's guaranteed never to happen.

File: src/mongo/MongoScrapbook.ts (hunk 7 above)
Comment (EricJizbaMSFT) [reply src/mongo/MongoScrapbook.ts]: see other comment near executeCommand

File: src/mongo/MongoScrapbook.ts [hunk 9]
```diff
@@ -112,16 +112,13 @@ async function executeCommand(activeEditor: vscode.TextEditor, database: MongoDa
 			throw new Error(`Error near line ${err.range.start.line}, column ${err.range.start.character}: '${err.message}'. Please check syntax.`);
 		}
 
-		if (command.name === 'find') {
+		if (command.name === 'find' && !!command.collection && !command.sendToShell) {
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoScrapbook.ts]: Can we move this code directly into `database.executeCommand`

```
if (command.name === 'find' && !!command.collection && !command.sendToShell) {
	await editorManager.showDocument(context, new MongoFindResultEditor(database, command), 'cosmos-result.json', { showInNextColumn: true });
} else {
	const result = await database.executeCommand(command, context);
	if (command.name === 'findOne' && !!command.collection) {
		await editorManager.showDocument(context, new MongoFindOneResultEditor(database, command.collection, result), 'cosmos-result.json', { showInNextColumn: true });
	} else { // failover also includes chained commands
		await vscodeUtil.showNewFile(result, 'result', '.json', activeEditor.viewColumn + 1);
		await refreshTreeAfterCommand(database, command, context);
	}
}
```

I just think there's a lot of duplicate checks here and I'm concerned we're missing a case of `command.sendToShell`. If you move it all into `database.executeCommand` and integrate it correctly, I think you'll only ever have to check `command.sendToShell` once

**Comment by EricJizbaMSFT** [comment]
Closing due to lack of activity. Feel free to re-open once you're actively working on this
